diff --git i/src/backend/access/transam/timeline.c w/src/backend/access/transam/timeline.c
index d80c8ffe0a7..648bfefe751 100644
--- i/src/backend/access/transam/timeline.c
+++ w/src/backend/access/transam/timeline.c
@@ -43,6 +43,10 @@
 #include "storage/fd.h"
 #include "utils/wait_event.h"
 
+// @czepes hash
+#include "catalog/pg_control.h"
+// @sepezc
+
 /*
  * Copies all timeline history files with id's between 'begin' and 'end'
  * from archive to pg_wal.
@@ -132,6 +136,9 @@ readTimeLineHistory(TimeLineID targetTLI)
 		uint32		switchpoint_hi;
 		uint32		switchpoint_lo;
 		int			nfields;
+		// @czepes hash
+		uint64 		timeline_hash;
+		// @sepezc
 
 		pgstat_report_wait_start(WAIT_EVENT_TIMELINE_HISTORY_READ);
 		res = fgets(fline, sizeof(fline), fd);
@@ -155,7 +162,15 @@ readTimeLineHistory(TimeLineID targetTLI)
 		if (*ptr == '\0' || *ptr == '#')
 			continue;
 
-		nfields = sscanf(fline, "%u\t%X/%08X", &tli, &switchpoint_hi, &switchpoint_lo);
+		// nfields = sscanf(fline, "%u\t%X/%08X", &tli, &switchpoint_hi, &switchpoint_lo);
+		// @czepes hash
+		nfields = sscanf(fline,
+						 "%u\t%X/%08X\t" UINT64_HEX_FORMAT,
+						 &tli,
+						 &switchpoint_hi,
+						 &switchpoint_lo,
+						 &timeline_hash);
+		// @sepezc
 
 		if (nfields < 1)
 		{
@@ -164,7 +179,10 @@ readTimeLineHistory(TimeLineID targetTLI)
 					(errmsg("syntax error in history file: %s", fline),
 					 errhint("Expected a numeric timeline ID.")));
 		}
-		if (nfields != 3)
+		// if (nfields != 3)
+		// @czepes hash
+		if (nfields < 3)
+		// @sepezc
 			ereport(FATAL,
 					(errmsg("syntax error in history file: %s", fline),
 					 errhint("Expected a write-ahead log switchpoint location.")));
@@ -181,6 +199,11 @@ readTimeLineHistory(TimeLineID targetTLI)
 		entry->begin = prevend;
 		entry->end = ((uint64) (switchpoint_hi)) << 32 | (uint64) switchpoint_lo;
 		prevend = entry->end;
+		// @czepes hash
+		entry->hash = nfields < 4
+			? TIMELINE_HASH_INITIAL_SEED
+			: timeline_hash;
+		// @sepezc
 
 		/* Build list with newest item first */
 		result = lcons(entry, result);
@@ -203,6 +226,9 @@ readTimeLineHistory(TimeLineID targetTLI)
 	entry->tli = targetTLI;
 	entry->begin = prevend;
 	entry->end = InvalidXLogRecPtr;
+	// @czepes hash
+	entry->hash = TIMELINE_HASH_INITIAL_SEED;
+	// @sepezc
 
 	result = lcons(entry, result);
 
@@ -302,8 +328,13 @@ findNewestTimeLine(TimeLineID startTLI)
  * emplacing a bogus file.
  */
 void
+// writeTimeLineHistory(TimeLineID newTLI, TimeLineID parentTLI,
+// 					 XLogRecPtr switchpoint, const char *reason)
+// @czepes hash
 writeTimeLineHistory(TimeLineID newTLI, TimeLineID parentTLI,
-					 XLogRecPtr switchpoint, const char *reason)
+					 XLogRecPtr switchpoint, const char *reason,
+					 uint64 timeline_hash)
+// @sepezc
 {
 	char		path[MAXPGPATH];
 	char		tmppath[MAXPGPATH];
@@ -399,12 +430,21 @@ writeTimeLineHistory(TimeLineID newTLI, TimeLineID parentTLI,
 	 * If we did have a parent file, insert an extra newline just in case the
 	 * parent file failed to end with one.
 	 */
+	// snprintf(buffer, sizeof(buffer),
+	// 		 "%s%u\t%X/%08X\t%s\n",
+	// 		 (srcfd < 0) ? "" : "\n",
+	// 		 parentTLI,
+	// 		 LSN_FORMAT_ARGS(switchpoint),
+	// 		 reason);
+	// @czepes hash
 	snprintf(buffer, sizeof(buffer),
-			 "%s%u\t%X/%08X\t%s\n",
+			 "%s%u\t%X/%08X\t" UINT64_HEX_FORMAT "\t%s",
 			 (srcfd < 0) ? "" : "\n",
 			 parentTLI,
 			 LSN_FORMAT_ARGS(switchpoint),
+			 timeline_hash,
 			 reason);
+	// @sepezc
 
 	nbytes = strlen(buffer);
 	errno = 0;
diff --git i/src/backend/access/transam/xlog.c w/src/backend/access/transam/xlog.c
index b23d8bbbdad..3c99fdf39ca 100644
--- i/src/backend/access/transam/xlog.c
+++ w/src/backend/access/transam/xlog.c
@@ -36,6 +36,7 @@
  *-------------------------------------------------------------------------
  */
 
+#include "c.h"
 #include "postgres.h"
 
 #include <ctype.h>
@@ -110,6 +111,12 @@
 #include "utils/varlena.h"
 #include "utils/wait_event.h"
 
+// @czepes hash
+#include "access/xlog.h"
+#include "access/xlogdefs.h"
+#include "common/hashfn.h"
+// @sepezc
+
 #ifdef WAL_DEBUG
 #include "utils/memutils.h"
 #endif
@@ -751,6 +758,11 @@ static void WALInsertLockUpdateInsertingAt(XLogRecPtr insertingAt);
 
 static void XLogChecksums(uint32 new_type);
 
+
+// @czepes hash
+static void XLogResetHashInternal(void);
+// @sepezc
+
 /*
  * Insert an XLOG record represented by an already-constructed chain of data
  * chunks.  This is a low-level routine; to construct the WAL record header
@@ -2510,6 +2522,11 @@ XLogWrite(XLogwrtRqst WriteRqst, TimeLineID tli, bool flexible)
 				XLogCtl->lastSegSwitchTime = (pg_time_t) time(NULL);
 				XLogCtl->lastSegSwitchLSN = LogwrtResult.Flush;
 
+				// @czepes hash
+				/* Flush current timeline's hash to pg_control file due to the end of segment. */
+				XLogUpdateHash(openLogSegNo, tli);
+				// @sepezc
+
 				/*
 				 * Request a checkpoint if we've consumed too much xlog since
 				 * the last one.  For speed, we first check using the local
@@ -4283,6 +4300,11 @@ InitControlFile(uint64 sysidentifier, uint32 data_checksum_version)
 	ControlFile->wal_log_hints = wal_log_hints;
 	ControlFile->track_commit_timestamp = track_commit_timestamp;
 	ControlFile->data_checksum_version = data_checksum_version;
+	// @czepes hash
+	ControlFile->timeline_hash = TIMELINE_HASH_INITIAL_SEED;
+	ControlFile->hash_tli = 0;
+	ControlFile->hash_segno = 0;
+	// @sepezc
 
 	/*
 	 * Set the data_checksum_version value into XLogCtl, which is where all
@@ -5627,7 +5649,10 @@ str_time(pg_time_t tnow, char *buf, size_t bufsize)
  * Initialize the first WAL segment on new timeline.
  */
 static void
-XLogInitNewTimeline(TimeLineID endTLI, XLogRecPtr endOfLog, TimeLineID newTLI)
+// XLogInitNewTimeline(TimeLineID endTLI, XLogRecPtr endOfLog, TimeLineID newTLI)
+// @czepes hash
+XLogInitNewTimeline(TimeLineID endTLI, XLogRecPtr endOfLog, TimeLineID newTLI, uint64 *end_tl_hash)
+// @sepezc
 {
 	char		xlogfname[MAXFNAMELEN];
 	XLogSegNo	endLogSegNo;
@@ -5696,6 +5721,19 @@ XLogInitNewTimeline(TimeLineID endTLI, XLogRecPtr endOfLog, TimeLineID newTLI)
 	 */
 	XLogFileName(xlogfname, newTLI, startLogSegNo, wal_segment_size);
 	XLogArchiveCleanup(xlogfname);
+
+	// @czepes hash
+	/*
+	 * Get updated ending timeline's hash and reset new timeline's hash.
+	 * Ending timeline's hash is written to new history file by writeTimeLineHistory().
+	 */
+	XLogUpdateHash(endLogSegNo, endTLI);
+
+	LWLockAcquire(ControlFileLock, LW_EXCLUSIVE);
+	*end_tl_hash = ControlFile->timeline_hash;
+	XLogResetHashInternal();
+	LWLockRelease(ControlFileLock);
+	// @sepezc
 }
 
 /*
@@ -5860,6 +5898,9 @@ StartupXLOG(void)
 	TransactionId oldestActiveXID;
 	bool		promoted = false;
 	char		timebuf[128];
+	// @czepes hash
+	uint64		end_tl_hash;
+	// @sepezc
 
 	/*
 	 * We should have an aux process resource owner to use, and we should not
@@ -6385,7 +6426,10 @@ StartupXLOG(void)
 		 * have a copy of the last block of the old WAL in
 		 * endOfRecovery->lastPage; we will use that below.)
 		 */
-		XLogInitNewTimeline(EndOfLogTLI, EndOfLog, newTLI);
+		// XLogInitNewTimeline(EndOfLogTLI, EndOfLog, newTLI);
+		// @czepes hash
+		XLogInitNewTimeline(EndOfLogTLI, EndOfLog, newTLI, &end_tl_hash);
+		// @sepezc
 
 		/*
 		 * Remove the signal files out of the way, so that we don't
@@ -6407,8 +6451,13 @@ StartupXLOG(void)
 		 * To minimize the window for that, try to do as little as possible
 		 * between here and writing the end-of-recovery record.
 		 */
+		// writeTimeLineHistory(newTLI, recoveryTargetTLI,
+		// 					 EndOfLog, endOfRecoveryInfo->recoveryStopReason);
+		// @czepes hash
 		writeTimeLineHistory(newTLI, recoveryTargetTLI,
-							 EndOfLog, endOfRecoveryInfo->recoveryStopReason);
+							 EndOfLog, endOfRecoveryInfo->recoveryStopReason,
+							 end_tl_hash);
+		// @sepezc
 
 		ereport(LOG,
 				(errmsg("archive recovery complete")));
@@ -10201,3 +10250,181 @@ SetWalWriterSleeping(bool sleeping)
 	XLogCtl->WalWriterSleeping = sleeping;
 	SpinLockRelease(&XLogCtl->info_lck);
 }
+
+// @czepes hash
+/*
+ * Reset current timeline's hash state to initial.
+ * Must be called at every the timeline switch.
+ */
+void XLogResetHash()
+{
+	LWLockAcquire(ControlFileLock, LW_EXCLUSIVE);
+	XLogResetHashInternal();
+	LWLockRelease(ControlFileLock);
+}
+
+static void XLogResetHashInternal(void)
+{
+	ControlFile->timeline_hash = TIMELINE_HASH_INITIAL_SEED;
+	UpdateControlFile();
+}
+
+/*
+ * Update current timeline's hash state with given 'tli+segno'
+ * segment's hash. Calls XLogHashSegment() to calculate hash.
+ *
+ * Returns true if any pages were hashed,
+ * false if pg_control contains hash for "future" segment
+ * or segment is empty and no pages were hashed.
+ */
+bool XLogUpdateHash(XLogSegNo segno, TimeLineID tli)
+{
+	TimeLineID	hash_tli;
+	XLogSegNo	hash_segno;
+	uint64		old_hash;
+	uint64		new_hash;
+
+	LWLockAcquire(ControlFileLock, LW_SHARED);
+	old_hash = ControlFile->timeline_hash;
+	hash_tli = ControlFile->hash_tli;
+	hash_segno = ControlFile->hash_segno;
+	LWLockRelease(ControlFileLock);
+
+	if (tli < hash_tli) {
+		return false;
+	}
+
+	if (tli == hash_tli && segno <= hash_segno) {
+		return false;
+	}
+
+	new_hash = XLogHashSegment(segno, tli, old_hash);
+
+	if (new_hash == old_hash) {
+		return false;
+	}
+
+	ereport(LOG,
+			errmsg("wal hash update: assigning new hash "
+			UINT64_HEX_FORMAT, new_hash));
+
+	LWLockAcquire(ControlFileLock, LW_EXCLUSIVE);
+	ControlFile->timeline_hash = new_hash;
+	ControlFile->hash_tli = tli;
+	ControlFile->hash_segno = segno;
+	UpdateControlFile();
+	LWLockRelease(ControlFileLock);
+
+	return true;
+}
+
+/*
+ * Calculate segments's hash by reading given 'tli+segno' segment
+ * page-by-page and calculating chained hash for each valid (non-torn) page.
+ * Stop on first non-valid (torn) page or at the end of segment.
+ *
+ * Returns calculated hash.
+ * It equals to TIMELINE_INITAL_HASH_SEED if no calculation was made.
+ */
+uint64 XLogHashSegment(XLogSegNo segno, TimeLineID tli, uint64 seed)
+{
+	char		path[MAXPGPATH];
+	int			fd;
+	Size		page_offset;
+	uint64		hash = seed;
+
+	ereport(LOG,
+			errmsg("wal hash: begin hashing for segment "
+				   UINT64_FORMAT, segno));
+
+	XLogFilePath(path, tli, segno, wal_segment_size);
+	fd = OpenTransientFile(path, O_RDONLY | PG_BINARY);
+
+	if (fd < 0) {
+		ereport(WARNING, errmsg("wal hash: failed to open segment %s", path));
+		return hash;
+	}
+
+	for (page_offset = 0;
+		 page_offset < wal_segment_size;
+		 page_offset += XLOG_BLCKSZ)
+	{
+		char			page[XLOG_BLCKSZ];
+		XLogPageHeader	header = (XLogPageHeader) page;
+		XLogRecPtr		pageaddr;
+		Size			nread;
+		uint64			new_hash = TIMELINE_HASH_INITIAL_SEED;
+
+		pageaddr = segno * wal_segment_size + page_offset;
+
+		nread = read(fd, page, XLOG_BLCKSZ);
+		if (nread != XLOG_BLCKSZ)
+		{
+			ereport(LOG,
+					errmsg("wal hash: failed to read full page at "
+						   "%X/%08X : %zu != %zu",
+						   LSN_FORMAT_ARGS(pageaddr), nread, XLOG_BLCKSZ));
+			break;
+		}
+
+		if (header->xlp_magic != XLOG_PAGE_MAGIC)
+		{
+			ereport(LOG,
+					errmsg("wal hash: magic mismatch at "
+						   "%X/%08X : %u != %u",
+						   LSN_FORMAT_ARGS(pageaddr), header->xlp_magic, XLOG_PAGE_MAGIC));
+			break;
+		}
+
+		if (header->xlp_pageaddr != pageaddr)
+		{
+			ereport(LOG,
+						  errmsg("wal hash: page address mismatch at "
+						   "%X/%08X : %X/%08X != %X/%08X",
+						   LSN_FORMAT_ARGS(pageaddr),
+						   LSN_FORMAT_ARGS(header->xlp_pageaddr),
+						   LSN_FORMAT_ARGS(pageaddr)));
+			break;
+		}
+
+		if (header->xlp_tli != tli)
+		{
+			ereport(LOG,
+					errmsg("wal hash: skipping page from other TLI at "
+						   "%X/%08X : %u != %u",
+						   LSN_FORMAT_ARGS(pageaddr), header->xlp_tli, tli));
+			continue;
+		}
+
+		while (new_hash == TIMELINE_HASH_INITIAL_SEED)
+		{
+			new_hash =
+				hash_bytes_extended((const unsigned char *) page,
+									XLOG_BLCKSZ,
+									hash);
+		}
+		hash = new_hash;
+	}
+
+	CloseTransientFile(fd);
+	return hash;
+}
+
+/* Get current timeline's hash from pg_control. */
+uint64 GetCurrentTimelineHash(TLHashState *state)
+{
+	uint64		timeline_hash;
+
+	LWLockAcquire(ControlFileLock, LW_SHARED);
+	timeline_hash = ControlFile->timeline_hash;
+	if (state != NULL)
+	{
+		state->hash = timeline_hash;
+		state->tli = ControlFile->hash_tli;
+		state->segno = ControlFile->hash_segno;
+	}
+	LWLockRelease(ControlFileLock);
+
+	return timeline_hash;
+}
+// @sepezc
diff --git i/src/backend/access/transam/xlogrecovery.c w/src/backend/access/transam/xlogrecovery.c
index 6de13b91748..dcf954477a1 100644
--- i/src/backend/access/transam/xlogrecovery.c
+++ w/src/backend/access/transam/xlogrecovery.c
@@ -69,6 +69,10 @@
 #include "utils/pg_rusage.h"
 #include "utils/wait_event.h"
 
+// @czepes hash
+#include "access/xlog.h"
+// @sepezc
+
 /* Unsupported old recovery command file names (relative to $PGDATA) */
 #define RECOVERY_COMMAND_FILE	"recovery.conf"
 #define RECOVERY_COMMAND_DONE	"recovery.done"
@@ -1628,6 +1632,14 @@ PerformWalRecovery(void)
 	XLogRecord *record;
 	bool		reachedRecoveryTarget = false;
 	TimeLineID	replayTLI;
+	// @czepes hash
+	/* Needed for segment switch catching and rehashig */
+	TimeLineID	prev_hash_tli;
+	XLogSegNo	prev_hash_segno;
+
+	XLByteToPrevSeg(RedoStartLSN, prev_hash_segno, wal_segment_size);
+	prev_hash_tli = RedoStartTLI;
+	// @sepezc
 
 	/*
 	 * Initialize shared variables for tracking progress of WAL replay, as if
@@ -1795,6 +1807,32 @@ PerformWalRecovery(void)
 			 */
 			ApplyWalRecord(xlogreader, record, &replayTLI);
 
+			// @czepes hash
+			/*
+			 * Identify if recovery process moved to the next segment.
+			 * If it did, try to hash this segment; it won't be hashed
+			 * if pg_control already has hash of "future" segment.
+			 */
+			{
+				XLogSegNo	cur_segno;
+				XLByteToPrevSeg(xlogreader->ReadRecPtr, cur_segno, wal_segment_size);
+
+				while (prev_hash_tli != replayTLI || prev_hash_segno != cur_segno)
+				{
+					XLogUpdateHash(prev_hash_segno, prev_hash_tli);
+
+					if (prev_hash_tli != replayTLI)
+					{
+						XLogResetHash();
+
+						prev_hash_tli = replayTLI;
+					}
+					else
+						prev_hash_segno++;
+				}
+			}
+			// @sepezc
+
 			/*
 			 * Wake up processes waiting for standby replay, write, or flush
 			 * LSN to reach current replay position.  Replay implies that the
diff --git i/src/backend/utils/misc/pg_controldata.c w/src/backend/utils/misc/pg_controldata.c
index d229ae35209..e9975adb3ae 100644
--- i/src/backend/utils/misc/pg_controldata.c
+++ w/src/backend/utils/misc/pg_controldata.c
@@ -28,6 +28,10 @@
 #include "utils/pg_lsn.h"
 #include "utils/timestamp.h"
 
+// @czepes hash
+#include "access/xlogdefs.h"
+// @sepezc
+
 Datum
 pg_control_system(PG_FUNCTION_ARGS)
 {
@@ -264,3 +268,39 @@ pg_control_init(PG_FUNCTION_ARGS)
 
 	PG_RETURN_DATUM(HeapTupleGetDatum(htup));
 }
+
+// @czepes hash
+Datum
+pg_control_hash(PG_FUNCTION_ARGS)
+{
+	Datum		values[3];
+	bool		nulls[3];
+	TupleDesc	tupdesc;
+	HeapTuple	htup;
+	TimeLineID	hash_tli;
+	XLogSegNo	hash_segno;
+	char		buf[16 + 1];
+	TLHashState	hash_state;
+
+	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+		elog(ERROR, "return type must be a row type");
+
+	GetCurrentTimelineHash(&hash_state);
+
+	snprintf(buf, sizeof(buf),
+			 UINT64_HEX_FORMAT,
+			 hash_state.hash);
+	values[0] = CStringGetTextDatum(buf);
+	nulls[0] = false;
+
+	values[1] = UInt32GetDatum(hash_state.tli);
+	nulls[1] = false;
+
+	values[2] = UInt64GetDatum(hash_state.segno);
+	nulls[2] = false;
+
+	htup = heap_form_tuple(tupdesc, values, nulls);
+
+	PG_RETURN_DATUM(HeapTupleGetDatum(htup));
+}
+// @sepezc
diff --git i/src/bin/pg_rewind/pg_rewind.c w/src/bin/pg_rewind/pg_rewind.c
index 2e86fd158d0..7eb53b2e10c 100644
--- i/src/bin/pg_rewind/pg_rewind.c
+++ w/src/bin/pg_rewind/pg_rewind.c
@@ -32,6 +32,10 @@
 #include "rewind_source.h"
 #include "storage/bufpage.h"
 
+// @czepes hash
+#include "access/xlogdefs.h"
+// @sepezc
+
 static void usage(const char *progname);
 
 static void perform_rewind(filemap_t *filemap, rewind_source *source,
@@ -56,6 +60,15 @@ static void findCommonAncestorTimeline(TimeLineHistoryEntry *a_history,
 static void ensureCleanShutdown(const char *argv0);
 static void disconnect_atexit(void);
 
+// @czepes hash
+static bool matchCurrentTimelines(TimeLineID source_tli,
+								  TimeLineID target_tli,
+								  TLHashState source_hash,
+								  TLHashState target_hash);
+static bool matchHashStates(TLHashState a, TLHashState b);
+static bool matchHashes(uint64 a, uint64 b);
+// @sepezc
+
 static ControlFileData ControlFile_target;
 static ControlFileData ControlFile_source;
 static ControlFileData ControlFile_source_after;
@@ -154,6 +167,10 @@ main(int argc, char **argv)
 	bool		rewind_needed;
 	bool		writerecoveryconf = false;
 	filemap_t  *filemap;
+	// @czepes hash
+	TLHashState	source_hash;
+	TLHashState	target_hash;
+	// @sepezc
 
 	pg_logging_init(argv[0]);
 	set_pglocale_pgservice(argv[0], PG_TEXTDOMAIN("pg_rewind"));
@@ -369,20 +386,42 @@ main(int argc, char **argv)
 	target_tli = Max(ControlFile_target.minRecoveryPointTLI,
 					 ControlFile_target.checkPointCopy.ThisTimeLineID);
 
+	// @czepes hash
+	source_hash = (TLHashState) {
+		.hash	= ControlFile_source.timeline_hash,
+		.tli	= ControlFile_source.hash_tli,
+		.segno	= ControlFile_source.hash_segno,
+	};
+	target_hash = (TLHashState) {
+		.hash	= ControlFile_target.timeline_hash,
+		.tli	= ControlFile_target.hash_tli,
+		.segno	= ControlFile_target.hash_segno,
+	};
+	// @sepezc
+
 	/*
 	 * Find the common ancestor timeline between the clusters.
 	 *
 	 * If both clusters are already on the same timeline, there's nothing to
 	 * do.
 	 */
-	if (target_tli == source_tli)
+	// if (target_tli == source_tli)
+	// {
+	// 	pg_log_info("source and target cluster are on the same timeline");
+	// 	rewind_needed = false;
+	// 	target_wal_endrec = InvalidXLogRecPtr;
+	// }
+	// @czepes hash
+	if (matchCurrentTimelines(source_tli, target_tli, source_hash, target_hash))
 	{
-		pg_log_info("source and target cluster are on the same timeline");
+		pg_log_info("source and target clusters have identical hashes");
 		rewind_needed = false;
 		target_wal_endrec = InvalidXLogRecPtr;
 	}
 	else
 	{
+		pg_log_info("history comparison is required to locate divergence location");
+	// @sepezc
 		XLogRecPtr	chkptendrec;
 		TimeLineHistoryEntry *sourceHistory;
 		int			sourceNentries;
@@ -862,6 +901,20 @@ MinXLogRecPtr(XLogRecPtr a, XLogRecPtr b)
 		return Min(a, b);
 }
 
+// @czepes hash
+/* Copy from test/regress/pg_regress.c */
+bool
+file_exists(const char *file)
+{
+	FILE	   *f = fopen(file, "r");
+
+	if (!f)
+		return false;
+	fclose(f);
+	return true;
+}
+// @sepezc
+
 /*
  * Retrieve timeline history for the source or target system.
  */
@@ -898,6 +951,13 @@ getTimelineHistory(TimeLineID tli, bool is_source, int *nentries)
 		pg_free(histfile);
 	}
 
+	// @czepes hash
+	/* Fetch current timeline's hash from pg_control */
+	(history + (*nentries) - 1)->hash = is_source
+		? ControlFile_source.timeline_hash
+		: ControlFile_target.timeline_hash;
+	// @sepezc
+
 	/* In debugging mode, print what we read */
 	if (debug)
 	{
@@ -913,9 +973,16 @@ getTimelineHistory(TimeLineID tli, bool is_source, int *nentries)
 			TimeLineHistoryEntry *entry;
 
 			entry = &history[i];
-			pg_log_debug("%u: %X/%08X - %X/%08X", entry->tli,
+			// pg_log_debug("%u: %X/%08X - %X/%08X", entry->tli,
+			// 			 LSN_FORMAT_ARGS(entry->begin),
+			// 			 LSN_FORMAT_ARGS(entry->end));
+			// @czepes hash
+			pg_log_debug("%u: %X/%08X - %X/%08X "
+						 UINT64_HEX_FORMAT, entry->tli,
 						 LSN_FORMAT_ARGS(entry->begin),
-						 LSN_FORMAT_ARGS(entry->end));
+						 LSN_FORMAT_ARGS(entry->end),
+						 entry->hash);
+			// @sepezc
 		}
 	}
 
@@ -935,6 +1002,9 @@ findCommonAncestorTimeline(TimeLineHistoryEntry *a_history, int a_nentries,
 {
 	int			i,
 				n;
+	// @czepes hash
+	bool		hash_mismatch = false;
+	// @sepezc
 
 	/*
 	 * Trace the history forward, until we hit the timeline diverge. It may
@@ -950,8 +1020,79 @@ findCommonAncestorTimeline(TimeLineHistoryEntry *a_history, int a_nentries,
 		if (a_history[i].tli != b_history[i].tli ||
 			a_history[i].begin != b_history[i].begin)
 			break;
+		// @czepes hash
+		/*
+		 * Hash mismatch on one of the timelines means that divergence
+		 * occured either:
+		 * - at the begin LSN for both of this timelines
+		 * - at the end LSN of one of this timelines
+		 *
+		 *       hash mistach at TLI 2
+		 *       v
+		 *   -1- | -2--------->
+		 *   -1- | -2- | -3--->
+		 *       ^     ^
+		 *       ^     or from here
+		 *       WAL may diverge from here
+		 *
+		 * We can't determine where the actual divergence is.
+		 *
+		 * This is the serious flaw of HASHing approach:
+		 *
+		 * With UUID: if UUIDs for timelines match we know beforehand
+		 * that this timelines MUST be the same and have identical WAL
+		 * at the begin LSN.
+		 *
+		 * With HASH: if HASHes for timelines match the only thing
+		 * we know is that this timeline WAL content is different
+		 * starting from some LSN. It may be different from the begin
+		 * LSN or at timeline switch of one of the clusters end LSN.
+		 *
+		 * What we can do here is either:
+		 * - walk through WAL and compare hashes for each segment,
+		 *   find first mismatch and start rewind from that segment;
+		 * - simply start rewind from the start of the timeline.
+		 *
+		 * Right now we do the latter and this cause issues:
+		 * Rewind algorithm always supposes that diverence point is
+		 * end LSN, hence walks through previous records to find
+		 * latest checkPoint.
+		 * If we set divergence as begin LSN, there may not be previous
+		 * records to look through => fatal error.
+		 */
+		if (!matchHashes(a_history[i].hash, b_history[i].hash))
+		{
+			hash_mismatch = true;
+			break;
+		}
+		// @sepezc
 	}
 
+	// @czepes hash
+	if (hash_mismatch)
+	{
+		if (i > 0)
+		{
+			pg_log_info("hash mismatch at TLI %d", i + 1);
+
+			*recptr = a_history[i].begin; /* == b[i].begin */
+			*tliIndex = i;
+			return;
+		}
+		/*
+		 * WARN:
+		 * We can get here hashes mismatch if at timeline 1.
+		 * We trust that WAL of timeline 1 on both clusters are identical
+		 * to the point of minimal end LSN and set it as divergence point.
+		 * If we set divergence point as begin LSN (which is InvalidXLogRecPtr)
+		 * we will get the pg_rewind will fail to find previous checkpoint and
+		 * will terminate with fatal error.
+		 * TODO: Somehow fix this?
+		 */
+		i++;
+	}
+	// @sepezc
+
 	if (i > 0)
 	{
 		i--;
@@ -1215,3 +1356,48 @@ disconnect_atexit(void)
 	if (conn != NULL)
 		PQfinish(conn);
 }
+// @czepes hash
+static bool matchHashes(uint64 a, uint64 b)
+{
+	if (a == TIMELINE_HASH_INITIAL_SEED
+		&& b == TIMELINE_HASH_INITIAL_SEED)
+		return false;
+
+	return a == b;
+}
+
+static bool matchHashStates(TLHashState a, TLHashState b)
+{
+	return matchHashes(a.hash, b.hash) &&
+			a.segno == b.segno && a.tli == b.tli;
+}
+
+static bool matchCurrentTimelines(TimeLineID source_tli,
+								  TimeLineID target_tli,
+								  TLHashState source_hash,
+								  TLHashState target_hash)
+{
+	if (target_tli != source_tli) {
+		return false;
+	}
+
+	pg_log_info("source and target cluster are on the same timeline");
+
+	if (target_tli == 1)
+	{
+		pg_log_info("source and target clusters are on timeline 1; nothing to rewind");
+		return true;
+	}
+
+	if (!matchHashStates(source_hash, target_hash))
+	{
+		pg_log_info("source and target clusters have different hashes" );
+		return false;
+	}
+	else
+	{
+		pg_log_info("source and target clusters have identical hashes");
+		return true;
+	}
+}
+// @sepezc
diff --git i/src/bin/pg_rewind/timeline.c w/src/bin/pg_rewind/timeline.c
index dda06eaa0bc..7df4bf1a0c2 100644
--- i/src/bin/pg_rewind/timeline.c
+++ w/src/bin/pg_rewind/timeline.c
@@ -11,6 +11,10 @@
 
 #include "access/timeline.h"
 #include "pg_rewind.h"
+// @czepes hash
+#include "catalog/pg_control.h"
+#include <time.h>
+// @sepezc
 
 /*
  * This is copy-pasted from the backend readTimeLineHistory, modified to
@@ -48,6 +52,9 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 		uint32		switchpoint_hi;
 		uint32		switchpoint_lo;
 		int			nfields;
+		// @czepes hash
+		uint64 		timeline_hash;
+		// @sepezc
 
 		fline = bufptr;
 		while (*bufptr && *bufptr != '\n')
@@ -66,7 +73,16 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 		if (*ptr == '\0' || *ptr == '#')
 			continue;
 
-		nfields = sscanf(fline, "%u\t%X/%08X", &tli, &switchpoint_hi, &switchpoint_lo);
+		// nfields = sscanf(fline, "%u\t%X/%08X", &tli, &switchpoint_hi, &switchpoint_lo);
+		// @czepes hash
+		nfields = sscanf(fline,
+						 "%u\t%X/%08X\t"
+						 UINT64_HEX_FORMAT,
+						 &tli,
+						 &switchpoint_hi,
+						 &switchpoint_lo,
+						 &timeline_hash);
+		// @sepezc
 
 		if (nfields < 1)
 		{
@@ -75,7 +91,10 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 			pg_log_error_detail("Expected a numeric timeline ID.");
 			exit(1);
 		}
-		if (nfields != 3)
+		// if (nfields != 3)
+		// @czepes hash
+		if (nfields < 3)
+		// @sepezc
 		{
 			pg_log_error("syntax error in history file: %s", fline);
 			pg_log_error_detail("Expected a write-ahead log switchpoint location.");
@@ -98,6 +117,11 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 		entry->begin = prevend;
 		entry->end = ((uint64) (switchpoint_hi)) << 32 | (uint64) switchpoint_lo;
 		prevend = entry->end;
+		// @czepes hash
+		entry->hash = nfields < 4
+			? TIMELINE_HASH_INITIAL_SEED
+			: timeline_hash;
+		// @sepezc
 
 		/* we ignore the remainder of each line */
 	}
@@ -123,6 +147,9 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 	entry->tli = targetTLI;
 	entry->begin = prevend;
 	entry->end = InvalidXLogRecPtr;
+	// @czepes hash
+	entry->hash = TIMELINE_HASH_INITIAL_SEED;
+	// @sepezc
 
 	*nentries = nlines;
 	return entries;
diff --git i/src/include/access/timeline.h w/src/include/access/timeline.h
index 3aee3419a5c..ed9ea7328f3 100644
--- i/src/include/access/timeline.h
+++ w/src/include/access/timeline.h
@@ -27,13 +27,21 @@ typedef struct
 	TimeLineID	tli;
 	XLogRecPtr	begin;			/* inclusive */
 	XLogRecPtr	end;			/* exclusive, InvalidXLogRecPtr means infinity */
+	// @czepes hash
+	uint64 		hash;			/* hash of TLI's last completed segment at switchpoint */
+	// @sepezc
 } TimeLineHistoryEntry;
 
 extern List *readTimeLineHistory(TimeLineID targetTLI);
 extern bool existsTimeLineHistory(TimeLineID probeTLI);
 extern TimeLineID findNewestTimeLine(TimeLineID startTLI);
+// extern void writeTimeLineHistory(TimeLineID newTLI, TimeLineID parentTLI,
+// 								 XLogRecPtr switchpoint, const char *reason);
+// @czepes hash
 extern void writeTimeLineHistory(TimeLineID newTLI, TimeLineID parentTLI,
-								 XLogRecPtr switchpoint, const char *reason);
+								 XLogRecPtr switchpoint, const char *reason,
+								 uint64 timeline_hash);
+// @sepezc
 extern void writeTimeLineHistoryFile(TimeLineID tli, const char *content, size_t size);
 extern void restoreTimeLineHistoryFiles(TimeLineID begin, TimeLineID end);
 extern bool tliInHistory(TimeLineID tli, List *expectedTLEs);
diff --git i/src/include/access/xlog.h w/src/include/access/xlog.h
index 338d68d7424..1820c36ea73 100644
--- i/src/include/access/xlog.h
+++ w/src/include/access/xlog.h
@@ -18,6 +18,9 @@
 #include "lib/stringinfo.h"
 #include "nodes/pg_list.h"
 
+// @czepes hash
+#include "catalog/pg_control.h" /* for TLHashState */
+// @sepezc
 
 /* Sync methods */
 enum WalSyncMethod
@@ -354,4 +357,11 @@ extern SessionBackupState get_backup_status(void);
 /* files to signal promotion to primary */
 #define PROMOTE_SIGNAL_FILE		"promote"
 
+// @czepes hash
+extern void XLogResetHash();
+extern bool XLogUpdateHash(XLogSegNo segno, TimeLineID tli);
+extern uint64 XLogHashSegment(XLogSegNo segno, TimeLineID tli, uint64 seed);
+uint64 GetCurrentTimelineHash(TLHashState *state);
+// @sepezc
+
 #endif							/* XLOG_H */
diff --git i/src/include/c.h w/src/include/c.h
index 20cfbac54e7..927039287c9 100644
--- i/src/include/c.h
+++ w/src/include/c.h
@@ -694,6 +694,10 @@ typedef uint64_t uint64;
 #define UINT64_FORMAT "%" PRIu64
 #define OID8_FORMAT "%" PRIu64
 
+// @czepes hash
+#define UINT64_HEX_FORMAT "%016" PRIx64
+// @sepezc
+
 /*
  * 128-bit signed and unsigned integers
  *		There currently is only limited support for such types.
diff --git i/src/include/catalog/pg_control.h w/src/include/catalog/pg_control.h
index 80b3a730e03..bb689cc570c 100644
--- i/src/include/catalog/pg_control.h
+++ w/src/include/catalog/pg_control.h
@@ -22,7 +22,12 @@
 
 
 /* Version identifier for this pg_control format */
-#define PG_CONTROL_VERSION	1902
+// #define PG_CONTROL_VERSION	1902
+// @czepes hash
+#define PG_CONTROL_VERSION	1903
+
+#define TIMELINE_HASH_INITIAL_SEED	UINT64CONST(0)
+// sepezc@
 
 /* Nonce key length, see below */
 #define MOCK_AUTH_NONCE_LEN		32
@@ -244,10 +249,35 @@ typedef struct ControlFileData
 	 */
 	char		mock_authentication_nonce[MOCK_AUTH_NONCE_LEN];
 
+	// @czepes hash
+	/*
+	 * NOTE:
+	 * timeline_hash is a chained hash of WAL segment bytes in current timeline.
+	 * It is consistent across nodes that receive identical WAL stream
+	 * (replication or archive). It is NOT guaranteed to match across
+	 * independent primary nodes, even if they execute similar SQL queries,
+	 * because that execution is non-deterministic.
+	 *
+	 * timeline_hash initialized to TIMELINE_HASH_INITIAL_SEED.
+	 * hash_tli + hash_segno represent currently hashed segment file.
+	 */
+	uint64		timeline_hash;
+	TimeLineID	hash_tli;
+	XLogSegNo	hash_segno;
+	// sepezc@
+
 	/* CRC of all above ... MUST BE LAST! */
 	pg_crc32c	crc;
 } ControlFileData;
 
+// @czepes hash
+typedef struct {
+	uint64		hash;
+	TimeLineID	tli;
+	XLogSegNo	segno;
+} TLHashState;
+// @sepezc
+
 /*
  * Maximum safe value of sizeof(ControlFileData).  For reliability's sake,
  * it's critical that pg_control updates be atomic writes.  That generally
diff --git i/src/include/catalog/pg_proc.dat w/src/include/catalog/pg_proc.dat
index 66c3c9a04cf..9f5400e5e55 100644
--- i/src/include/catalog/pg_proc.dat
+++ w/src/include/catalog/pg_proc.dat
@@ -12454,6 +12454,17 @@
   proargnames => '{max_data_alignment,database_block_size,blocks_per_segment,wal_block_size,bytes_per_wal_segment,max_identifier_length,max_index_columns,max_toast_chunk_size,large_object_chunk_size,float8_pass_by_value,data_page_checksum_version,default_char_signedness}',
   prosrc => 'pg_control_init' },
 
+# @czepes hash
+{ oid => '9286',
+  descr => 'pg_controldata hash state information as a function',
+  proname => 'pg_control_hash', provolatile => 'v', prorettype => 'record',
+  proargtypes => '',
+  proallargtypes => '{text,int4,int8}',
+  proargmodes => '{o,o,o}',
+  proargnames => '{timeline_hash,hash_tli,hash_seg}',
+  prosrc => 'pg_control_hash' },
+# @sepezc
+
 # subscripting support for built-in types
 { oid => '6179', descr => 'standard array subscripting support',
   proname => 'array_subscript_handler',
