diff --git a/src/backend/access/transam/timeline.c b/src/backend/access/transam/timeline.c
index d80c8ff..3c114a7 100644
--- a/src/backend/access/transam/timeline.c
+++ b/src/backend/access/transam/timeline.c
@@ -43,6 +43,8 @@
 #include "storage/fd.h"
 #include "utils/wait_event.h"
 
+#include "access/timeline_hash.h"
+
 /*
  * Copies all timeline history files with id's between 'begin' and 'end'
  * from archive to pg_wal.
@@ -132,6 +134,7 @@ readTimeLineHistory(TimeLineID targetTLI)
 		uint32		switchpoint_hi;
 		uint32		switchpoint_lo;
 		int			nfields;
+		TLHashState hstate;
 
 		pgstat_report_wait_start(WAIT_EVENT_TIMELINE_HISTORY_READ);
 		res = fgets(fline, sizeof(fline), fd);
@@ -155,7 +158,12 @@ readTimeLineHistory(TimeLineID targetTLI)
 		if (*ptr == '\0' || *ptr == '#')
 			continue;
 
-		nfields = sscanf(fline, "%u\t%X/%08X", &tli, &switchpoint_hi, &switchpoint_lo);
+		nfields = sscanf(fline,
+						 "%u\t%X/%08X\t" TLHASHST_FORMAT,
+						 &tli,
+						 &switchpoint_hi,
+						 &switchpoint_lo,
+						 TLHASHST_SCANF_ARGS(hstate));
 
 		if (nfields < 1)
 		{
@@ -164,7 +172,7 @@ readTimeLineHistory(TimeLineID targetTLI)
 					(errmsg("syntax error in history file: %s", fline),
 					 errhint("Expected a numeric timeline ID.")));
 		}
-		if (nfields != 3)
+		if (nfields < 3)
 			ereport(FATAL,
 					(errmsg("syntax error in history file: %s", fline),
 					 errhint("Expected a write-ahead log switchpoint location.")));
@@ -181,6 +189,10 @@ readTimeLineHistory(TimeLineID targetTLI)
 		entry->begin = prevend;
 		entry->end = ((uint64) (switchpoint_hi)) << 32 | (uint64) switchpoint_lo;
 		prevend = entry->end;
+		entry->hstate =
+			nfields < 3 + TLHASHST_FIELDS
+			? TLHASHST_INITIAL
+			: hstate;
 
 		/* Build list with newest item first */
 		result = lcons(entry, result);
@@ -203,6 +215,7 @@ readTimeLineHistory(TimeLineID targetTLI)
 	entry->tli = targetTLI;
 	entry->begin = prevend;
 	entry->end = InvalidXLogRecPtr;
+	entry->hstate = TLHASHST_INITIAL;
 
 	result = lcons(entry, result);
 
@@ -303,7 +316,8 @@ findNewestTimeLine(TimeLineID startTLI)
  */
 void
 writeTimeLineHistory(TimeLineID newTLI, TimeLineID parentTLI,
-					 XLogRecPtr switchpoint, const char *reason)
+					 XLogRecPtr switchpoint, const char *reason,
+					 TLHashState *hstate)
 {
 	char		path[MAXPGPATH];
 	char		tmppath[MAXPGPATH];
@@ -400,10 +414,11 @@ writeTimeLineHistory(TimeLineID newTLI, TimeLineID parentTLI,
 	 * parent file failed to end with one.
 	 */
 	snprintf(buffer, sizeof(buffer),
-			 "%s%u\t%X/%08X\t%s\n",
+			 "%s%u\t%X/%08X\t" TLHASHST_FORMAT "\t%s",
 			 (srcfd < 0) ? "" : "\n",
 			 parentTLI,
 			 LSN_FORMAT_ARGS(switchpoint),
+			 TLHASHST_PRINT_ARGS((*hstate)),
 			 reason);
 
 	nbytes = strlen(buffer);
diff --git a/src/backend/access/transam/xlog.c b/src/backend/access/transam/xlog.c
index b23d8bb..8329ee1 100644
--- a/src/backend/access/transam/xlog.c
+++ b/src/backend/access/transam/xlog.c
@@ -110,6 +110,8 @@
 #include "utils/varlena.h"
 #include "utils/wait_event.h"
 
+#include "common/hashfn.h" /* for hash_bytes_extended() */
+
 #ifdef WAL_DEBUG
 #include "utils/memutils.h"
 #endif
@@ -751,6 +753,8 @@ static void WALInsertLockUpdateInsertingAt(XLogRecPtr insertingAt);
 
 static void XLogChecksums(uint32 new_type);
 
+static void XLogResetHashInternal(void);
+
 /*
  * Insert an XLOG record represented by an already-constructed chain of data
  * chunks.  This is a low-level routine; to construct the WAL record header
@@ -2510,6 +2514,9 @@ XLogWrite(XLogwrtRqst WriteRqst, TimeLineID tli, bool flexible)
 				XLogCtl->lastSegSwitchTime = (pg_time_t) time(NULL);
 				XLogCtl->lastSegSwitchLSN = LogwrtResult.Flush;
 
+				/* Flush current timeline's hash to pg_control file due to the end of segment. */
+				XLogUpdateCurrentHash(openLogSegNo, tli);
+
 				/*
 				 * Request a checkpoint if we've consumed too much xlog since
 				 * the last one.  For speed, we first check using the local
@@ -4283,6 +4290,7 @@ InitControlFile(uint64 sysidentifier, uint32 data_checksum_version)
 	ControlFile->wal_log_hints = wal_log_hints;
 	ControlFile->track_commit_timestamp = track_commit_timestamp;
 	ControlFile->data_checksum_version = data_checksum_version;
+	ControlFile->hstate = TLHASHST_INITIAL;
 
 	/*
 	 * Set the data_checksum_version value into XLogCtl, which is where all
@@ -5627,7 +5635,7 @@ str_time(pg_time_t tnow, char *buf, size_t bufsize)
  * Initialize the first WAL segment on new timeline.
  */
 static void
-XLogInitNewTimeline(TimeLineID endTLI, XLogRecPtr endOfLog, TimeLineID newTLI)
+XLogInitNewTimeline(TimeLineID endTLI, XLogRecPtr endOfLog, TimeLineID newTLI, TLHashState *end_tl_hash)
 {
 	char		xlogfname[MAXFNAMELEN];
 	XLogSegNo	endLogSegNo;
@@ -5696,6 +5704,17 @@ XLogInitNewTimeline(TimeLineID endTLI, XLogRecPtr endOfLog, TimeLineID newTLI)
 	 */
 	XLogFileName(xlogfname, newTLI, startLogSegNo, wal_segment_size);
 	XLogArchiveCleanup(xlogfname);
+
+	/*
+	 * Get updated ending timeline's hash and reset new timeline's hash.
+	 * Ending timeline's hash is written to new history file by writeTimeLineHistory().
+	 */
+	XLogUpdateCurrentHash(endLogSegNo, endTLI);
+
+	LWLockAcquire(ControlFileLock, LW_EXCLUSIVE);
+	*end_tl_hash = ControlFile->hstate;
+	XLogResetHashInternal();
+	LWLockRelease(ControlFileLock);
 }
 
 /*
@@ -5860,6 +5879,7 @@ StartupXLOG(void)
 	TransactionId oldestActiveXID;
 	bool		promoted = false;
 	char		timebuf[128];
+	TLHashState	end_hstate;
 
 	/*
 	 * We should have an aux process resource owner to use, and we should not
@@ -6385,7 +6405,7 @@ StartupXLOG(void)
 		 * have a copy of the last block of the old WAL in
 		 * endOfRecovery->lastPage; we will use that below.)
 		 */
-		XLogInitNewTimeline(EndOfLogTLI, EndOfLog, newTLI);
+		XLogInitNewTimeline(EndOfLogTLI, EndOfLog, newTLI, &end_hstate);
 
 		/*
 		 * Remove the signal files out of the way, so that we don't
@@ -6408,7 +6428,8 @@ StartupXLOG(void)
 		 * between here and writing the end-of-recovery record.
 		 */
 		writeTimeLineHistory(newTLI, recoveryTargetTLI,
-							 EndOfLog, endOfRecoveryInfo->recoveryStopReason);
+							 EndOfLog, endOfRecoveryInfo->recoveryStopReason,
+							 &end_hstate);
 
 		ereport(LOG,
 				(errmsg("archive recovery complete")));
@@ -10201,3 +10222,191 @@ SetWalWriterSleeping(bool sleeping)
 	XLogCtl->WalWriterSleeping = sleeping;
 	SpinLockRelease(&XLogCtl->info_lck);
 }
+
+bool XLogHashIsSet(TLHashState hstate)
+{
+	return hstate.flags & TLHASHST_INIT_FLAG;
+}
+
+/*
+ * Reset current timeline's hash state to initial.
+ * Must be called at every the timeline switch.
+ */
+void XLogResetCurrentHash()
+{
+	LWLockAcquire(ControlFileLock, LW_EXCLUSIVE);
+	XLogResetHashInternal();
+	LWLockRelease(ControlFileLock);
+}
+
+static void XLogResetHashInternal(void)
+{
+	ControlFile->hstate.flags &= ~TLHASHST_INIT_FLAG;
+	UpdateControlFile();
+}
+
+/*
+ * Update current timeline's hash state with given 'tli+segno'
+ * segment's hash. Calls XLogHashSegment() to calculate hash.
+ *
+ * Returns true if any pages were hashed,
+ * false if pg_control contains hash for "future" segment
+ * or segment is empty and no pages were hashed.
+ */
+bool XLogUpdateCurrentHash(XLogSegNo segno, TimeLineID tli)
+{
+	TLHashState	hstate;
+	uint64		new_hash;
+
+	GetCurrentTimelineHash(&hstate);
+
+	if (tli < hstate.tli) {
+		return false;
+	}
+
+	if (tli == hstate.tli && segno <= hstate.segno) {
+		return false;
+	}
+
+	new_hash = XLogHashSegment(segno, tli, hstate.hash);
+
+	if (new_hash == hstate.hash) {
+		return false;
+	}
+
+	hstate.hash = new_hash;
+	hstate.segno = segno;
+	hstate.tli = tli;
+	hstate.flags |= TLHASHST_INIT_FLAG;
+
+	ereport(LOG,
+			errmsg("wal hash update: updating hash state "
+			TLHASHST_FORMAT, TLHASHST_PRINT_ARGS(hstate)));
+
+	SetCurrentTimelineHash(&hstate);
+
+	return true;
+}
+
+/*
+ * Calculate segments's hash by reading given 'tli+segno' segment
+ * page-by-page and calculating chained hash for each valid (non-torn) page.
+ * Stop on first non-valid (torn) page or at the end of segment.
+ *
+ * Returns calculated hash.
+ * It equals to 'seed' if no calculation was made.
+ */
+uint64 XLogHashSegment(XLogSegNo segno, TimeLineID tli, uint64 seed)
+{
+	char		path[MAXPGPATH];
+	int			fd;
+	Size		page_offset;
+	uint64		hash = seed;
+	TimeLineID	skipped;
+
+	ereport(LOG,
+			errmsg("wal hash: begin hashing for segment "
+				   UINT64_FORMAT, segno));
+
+	XLogFilePath(path, tli, segno, wal_segment_size);
+	fd = OpenTransientFile(path, O_RDONLY | PG_BINARY);
+
+	if (fd < 0) {
+		ereport(WARNING, errmsg("wal hash: failed to open segment %s", path));
+		return hash;
+	}
+
+	for (page_offset = 0;
+		 page_offset < wal_segment_size;
+		 page_offset += XLOG_BLCKSZ)
+	{
+		char			page[XLOG_BLCKSZ];
+		XLogPageHeader	header = (XLogPageHeader) page;
+		XLogRecPtr		pageaddr;
+		Size			nread;
+
+		pageaddr = segno * wal_segment_size + page_offset;
+
+		nread = read(fd, page, XLOG_BLCKSZ);
+		if (nread != XLOG_BLCKSZ)
+		{
+			ereport(LOG,
+					errmsg("wal hash: failed to read full page at "
+						   "%X/%08X : %zu != %zu",
+						   LSN_FORMAT_ARGS(pageaddr), nread, XLOG_BLCKSZ));
+			break;
+		}
+
+		if (header->xlp_magic != XLOG_PAGE_MAGIC)
+		{
+			ereport(LOG,
+					errmsg("wal hash: magic mismatch at "
+						   "%X/%08X : %u != %u",
+						   LSN_FORMAT_ARGS(pageaddr), header->xlp_magic, XLOG_PAGE_MAGIC));
+			break;
+		}
+
+		if (header->xlp_pageaddr != pageaddr)
+		{
+			ereport(LOG,
+						  errmsg("wal hash: page address mismatch at "
+						   "%X/%08X : %X/%08X != %X/%08X",
+						   LSN_FORMAT_ARGS(pageaddr),
+						   LSN_FORMAT_ARGS(header->xlp_pageaddr),
+						   LSN_FORMAT_ARGS(pageaddr)));
+			break;
+		}
+
+		if (header->xlp_tli != tli)
+		{
+			/* Print skip message only once per TLI */
+			if (skipped != header->xlp_tli)
+			{
+				skipped = header->xlp_tli;
+				ereport(LOG,
+						errmsg("wal hash: skipping page from other TLI at "
+							   "%X/%08X : %u != %u",
+							   LSN_FORMAT_ARGS(pageaddr), header->xlp_tli, tli));
+			}
+			continue;
+		}
+
+		hash = hash_bytes_extended((const unsigned char *) page,
+								   XLOG_BLCKSZ,
+								   hash);
+	}
+
+	CloseTransientFile(fd);
+	return hash;
+}
+
+/*
+ * Set and flush current timeline's hash state to pg_control.
+ */
+void SetCurrentTimelineHash(TLHashState *hstate)
+{
+	if (hstate == NULL) return;
+
+	LWLockAcquire(ControlFileLock, LW_EXCLUSIVE);
+	ControlFile->hstate = *hstate;
+	UpdateControlFile();
+	LWLockRelease(ControlFileLock);
+}
+
+/*
+ * Get current timeline's hash from pg_control.
+ */
+uint64 GetCurrentTimelineHash(TLHashState *hstate)
+{
+	uint64		timeline_hash;
+
+	LWLockAcquire(ControlFileLock, LW_SHARED);
+	timeline_hash = ControlFile->hstate.hash;
+	if (hstate != NULL)
+	{
+		*hstate = ControlFile->hstate;
+	}
+	LWLockRelease(ControlFileLock);
+
+	return timeline_hash;
+}
diff --git a/src/backend/access/transam/xlogrecovery.c b/src/backend/access/transam/xlogrecovery.c
index 6de13b9..1815589 100644
--- a/src/backend/access/transam/xlogrecovery.c
+++ b/src/backend/access/transam/xlogrecovery.c
@@ -1628,6 +1628,12 @@ PerformWalRecovery(void)
 	XLogRecord *record;
 	bool		reachedRecoveryTarget = false;
 	TimeLineID	replayTLI;
+	/* Needed for segment switch catching and rehashig */
+	TimeLineID	prev_hash_tli;
+	XLogSegNo	prev_hash_segno;
+
+	XLByteToPrevSeg(RedoStartLSN, prev_hash_segno, wal_segment_size);
+	prev_hash_tli = RedoStartTLI;
 
 	/*
 	 * Initialize shared variables for tracking progress of WAL replay, as if
@@ -1795,6 +1801,30 @@ PerformWalRecovery(void)
 			 */
 			ApplyWalRecord(xlogreader, record, &replayTLI);
 
+			/*
+			 * Identify if recovery process moved to the next segment.
+			 * If it did, try to hash this segment; it won't be hashed
+			 * if pg_control already has hash of "future" segment.
+			 */
+			{
+				XLogSegNo	cur_segno;
+				XLByteToPrevSeg(xlogreader->ReadRecPtr, cur_segno, wal_segment_size);
+
+				while (prev_hash_tli != replayTLI || prev_hash_segno != cur_segno)
+				{
+					XLogUpdateCurrentHash(prev_hash_segno, prev_hash_tli);
+
+					if (prev_hash_tli != replayTLI)
+					{
+						XLogResetCurrentHash();
+
+						prev_hash_tli = replayTLI;
+					}
+					else
+						prev_hash_segno++;
+				}
+			}
+
 			/*
 			 * Wake up processes waiting for standby replay, write, or flush
 			 * LSN to reach current replay position.  Replay implies that the
diff --git a/src/backend/utils/misc/pg_controldata.c b/src/backend/utils/misc/pg_controldata.c
index d229ae3..e980338 100644
--- a/src/backend/utils/misc/pg_controldata.c
+++ b/src/backend/utils/misc/pg_controldata.c
@@ -264,3 +264,38 @@ pg_control_init(PG_FUNCTION_ARGS)
 
 	PG_RETURN_DATUM(HeapTupleGetDatum(htup));
 }
+
+Datum
+pg_control_hash(PG_FUNCTION_ARGS)
+{
+	Datum		values[4];
+	bool		nulls[4];
+	TupleDesc	tupdesc;
+	HeapTuple	htup;
+	TLHashState	hstate;
+	char		buf[16 + 1];
+
+	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+		elog(ERROR, "return type must be a row type");
+
+	GetCurrentTimelineHash(&hstate);
+
+	snprintf(buf, sizeof(buf),
+			 UINT64_HEX_FORMAT,
+			 hstate.hash);
+	values[0] = CStringGetTextDatum(buf);
+	nulls[0] = false;
+
+	values[1] = UInt32GetDatum(hstate.tli);
+	nulls[1] = false;
+
+	values[2] = UInt64GetDatum(hstate.segno);
+	nulls[2] = false;
+
+	values[3] = BoolGetDatum(XLogHashIsSet(hstate));
+	nulls[3] = false;
+
+	htup = heap_form_tuple(tupdesc, values, nulls);
+
+	PG_RETURN_DATUM(HeapTupleGetDatum(htup));
+}
diff --git a/src/bin/pg_rewind/pg_rewind.c b/src/bin/pg_rewind/pg_rewind.c
index 2e86fd1..59f46cf 100644
--- a/src/bin/pg_rewind/pg_rewind.c
+++ b/src/bin/pg_rewind/pg_rewind.c
@@ -56,6 +56,12 @@ static void findCommonAncestorTimeline(TimeLineHistoryEntry *a_history,
 static void ensureCleanShutdown(const char *argv0);
 static void disconnect_atexit(void);
 
+static bool matchCurrentTimelines(TimeLineID source_tli,
+								  TimeLineID target_tli,
+								  TLHashState source_hash,
+								  TLHashState target_hash);
+static bool matchHashStates(TLHashState a, TLHashState b);
+
 static ControlFileData ControlFile_target;
 static ControlFileData ControlFile_source;
 static ControlFileData ControlFile_source_after;
@@ -154,6 +160,8 @@ main(int argc, char **argv)
 	bool		rewind_needed;
 	bool		writerecoveryconf = false;
 	filemap_t  *filemap;
+	TLHashState	source_hash;
+	TLHashState	target_hash;
 
 	pg_logging_init(argv[0]);
 	set_pglocale_pgservice(argv[0], PG_TEXTDOMAIN("pg_rewind"));
@@ -369,15 +377,18 @@ main(int argc, char **argv)
 	target_tli = Max(ControlFile_target.minRecoveryPointTLI,
 					 ControlFile_target.checkPointCopy.ThisTimeLineID);
 
+	source_hash = ControlFile_source.hstate;
+	target_hash = ControlFile_target.hstate;
+
 	/*
 	 * Find the common ancestor timeline between the clusters.
 	 *
 	 * If both clusters are already on the same timeline, there's nothing to
 	 * do.
 	 */
-	if (target_tli == source_tli)
+	if (matchCurrentTimelines(source_tli, target_tli, source_hash, target_hash))
 	{
-		pg_log_info("source and target cluster are on the same timeline");
+		pg_log_info("source and target clusters have identical hashes");
 		rewind_needed = false;
 		target_wal_endrec = InvalidXLogRecPtr;
 	}
@@ -387,6 +398,8 @@ main(int argc, char **argv)
 		TimeLineHistoryEntry *sourceHistory;
 		int			sourceNentries;
 
+		pg_log_info("history comparison is required to locate divergence location");
+
 		/*
 		 * Retrieve timelines for both source and target, and find the point
 		 * where they diverged.
@@ -898,6 +911,11 @@ getTimelineHistory(TimeLineID tli, bool is_source, int *nentries)
 		pg_free(histfile);
 	}
 
+	/* Fetch current timeline's hash from pg_control */
+	(history + (*nentries) - 1)->hstate = is_source
+		? ControlFile_source.hstate
+		: ControlFile_target.hstate;
+
 	/* In debugging mode, print what we read */
 	if (debug)
 	{
@@ -913,9 +931,11 @@ getTimelineHistory(TimeLineID tli, bool is_source, int *nentries)
 			TimeLineHistoryEntry *entry;
 
 			entry = &history[i];
-			pg_log_debug("%u: %X/%08X - %X/%08X", entry->tli,
+			pg_log_debug("%u: %X/%08X - %X/%08X "
+						 TLHASHST_FORMAT, entry->tli,
 						 LSN_FORMAT_ARGS(entry->begin),
-						 LSN_FORMAT_ARGS(entry->end));
+						 LSN_FORMAT_ARGS(entry->end),
+						 TLHASHST_PRINT_ARGS(entry->hstate));
 		}
 	}
 
@@ -935,6 +955,7 @@ findCommonAncestorTimeline(TimeLineHistoryEntry *a_history, int a_nentries,
 {
 	int			i,
 				n;
+	bool		hash_mismatch = false;
 
 	/*
 	 * Trace the history forward, until we hit the timeline diverge. It may
@@ -950,6 +971,74 @@ findCommonAncestorTimeline(TimeLineHistoryEntry *a_history, int a_nentries,
 		if (a_history[i].tli != b_history[i].tli ||
 			a_history[i].begin != b_history[i].begin)
 			break;
+
+		/*
+		 * Hash mismatch on one of the timelines means that divergence
+		 * occured either:
+		 * - at the begin LSN for both of this timelines
+		 * - at the end LSN of one of this timelines
+		 *
+		 *       hash mistach at TLI 2
+		 *       v
+		 *   -1- | -2--------->
+		 *   -1- | -2- | -3--->
+		 *       ^     ^
+		 *       ^     or from here
+		 *       WAL may diverge from here
+		 *
+		 * We can't determine where the actual divergence is.
+		 *
+		 * This is the serious flaw of HASHing approach:
+		 *
+		 * With UUID: if UUIDs for timelines match we know beforehand
+		 * that this timelines MUST be the same and have identical WAL
+		 * at the begin LSN.
+		 *
+		 * With HASH: if HASHes for timelines match the only thing
+		 * we know is that this timeline WAL content is different
+		 * starting from some LSN. It may be different from the begin
+		 * LSN or at timeline switch of one of the clusters end LSN.
+		 *
+		 * What we can do here is either:
+		 * - walk through WAL and compare hashes for each segment,
+		 *   find first mismatch and start rewind from that segment;
+		 * - simply start rewind from the start of the timeline.
+		 *
+		 * Right now we do the latter and this cause issues:
+		 * Rewind algorithm always supposes that diverence point is
+		 * end LSN, hence walks through previous records to find
+		 * latest checkPoint.
+		 * If we set divergence as begin LSN, there may not be previous
+		 * records to look through => fatal error.
+		 */
+		if (!matchHashStates(a_history[i].hstate, b_history[i].hstate))
+		{
+			hash_mismatch = true;
+			break;
+		}
+	}
+
+	if (hash_mismatch)
+	{
+		if (i > 0)
+		{
+			pg_log_info("hash mismatch at TLI %d", i + 1);
+
+			*recptr = a_history[i].begin; /* == b[i].begin */
+			*tliIndex = i;
+			return;
+		}
+		/*
+		 * WARN:
+		 * We can get here hashes mismatch if at timeline 1.
+		 * We trust that WAL of timeline 1 on both clusters are identical
+		 * to the point of minimal end LSN and set it as divergence point.
+		 * If we set divergence point as begin LSN (which is InvalidXLogRecPtr)
+		 * we will get the pg_rewind will fail to find previous checkpoint and
+		 * will terminate with fatal error.
+		 * TODO: Somehow fix this?
+		 */
+		i++;
 	}
 
 	if (i > 0)
@@ -1215,3 +1304,46 @@ disconnect_atexit(void)
 	if (conn != NULL)
 		PQfinish(conn);
 }
+
+static bool XLogHashIsSet(TLHashState hstate)
+{
+	return hstate.flags & TLHASHST_INIT_FLAG;
+}
+
+static bool matchHashStates(TLHashState a, TLHashState b)
+{
+	/* Unless both hashes are initialized, we cannot match them. */
+	if (!XLogHashIsSet(a) || !XLogHashIsSet(b))
+		return false;
+
+	return a.hash == b.hash && a.tli == b.tli && a.segno == b.segno;
+}
+
+static bool matchCurrentTimelines(TimeLineID source_tli,
+								  TimeLineID target_tli,
+								  TLHashState source_hash,
+								  TLHashState target_hash)
+{
+	if (target_tli != source_tli) {
+		return false;
+	}
+
+	pg_log_info("source and target cluster are on the same timeline");
+
+	if (target_tli == 1)
+	{
+		pg_log_info("source and target clusters are on timeline 1; nothing to rewind");
+		return true;
+	}
+
+	if (!matchHashStates(source_hash, target_hash))
+	{
+		pg_log_info("source and target clusters have different hashes" );
+		return false;
+	}
+	else
+	{
+		pg_log_info("source and target clusters have identical hashes");
+		return true;
+	}
+}
diff --git a/src/bin/pg_rewind/timeline.c b/src/bin/pg_rewind/timeline.c
index dda06ea..68c7ac2 100644
--- a/src/bin/pg_rewind/timeline.c
+++ b/src/bin/pg_rewind/timeline.c
@@ -12,6 +12,8 @@
 #include "access/timeline.h"
 #include "pg_rewind.h"
 
+#include "access/timeline_hash.h" /* for TIMELINE_HASH_INITIAL_SEED */
+
 /*
  * This is copy-pasted from the backend readTimeLineHistory, modified to
  * return a malloc'd array and to work without backend functions.
@@ -48,6 +50,7 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 		uint32		switchpoint_hi;
 		uint32		switchpoint_lo;
 		int			nfields;
+		TLHashState	hstate;
 
 		fline = bufptr;
 		while (*bufptr && *bufptr != '\n')
@@ -66,7 +69,12 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 		if (*ptr == '\0' || *ptr == '#')
 			continue;
 
-		nfields = sscanf(fline, "%u\t%X/%08X", &tli, &switchpoint_hi, &switchpoint_lo);
+		nfields = sscanf(fline,
+						 "%u\t%X/%08X\t" TLHASHST_FORMAT,
+						 &tli,
+						 &switchpoint_hi,
+						 &switchpoint_lo,
+						 TLHASHST_SCANF_ARGS(hstate));
 
 		if (nfields < 1)
 		{
@@ -75,7 +83,7 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 			pg_log_error_detail("Expected a numeric timeline ID.");
 			exit(1);
 		}
-		if (nfields != 3)
+		if (nfields < 3)
 		{
 			pg_log_error("syntax error in history file: %s", fline);
 			pg_log_error_detail("Expected a write-ahead log switchpoint location.");
@@ -98,6 +106,10 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 		entry->begin = prevend;
 		entry->end = ((uint64) (switchpoint_hi)) << 32 | (uint64) switchpoint_lo;
 		prevend = entry->end;
+		entry->hstate =
+			nfields < 3 + TLHASHST_FIELDS
+			? TLHASHST_INITIAL
+			: hstate;
 
 		/* we ignore the remainder of each line */
 	}
@@ -123,6 +135,7 @@ rewind_parseTimeLineHistory(char *buffer, TimeLineID targetTLI, int *nentries)
 	entry->tli = targetTLI;
 	entry->begin = prevend;
 	entry->end = InvalidXLogRecPtr;
+	entry->hstate = TLHASHST_INITIAL;
 
 	*nentries = nlines;
 	return entries;
diff --git a/src/include/access/timeline.h b/src/include/access/timeline.h
index 3aee341..3f24eed 100644
--- a/src/include/access/timeline.h
+++ b/src/include/access/timeline.h
@@ -14,6 +14,8 @@
 #include "access/xlogdefs.h"
 #include "nodes/pg_list.h"
 
+#include "access/timeline_hash.h"
+
 /*
  * A list of these structs describes the timeline history of the server. Each
  * TimeLineHistoryEntry represents a piece of WAL belonging to the history,
@@ -27,13 +29,15 @@ typedef struct
 	TimeLineID	tli;
 	XLogRecPtr	begin;			/* inclusive */
 	XLogRecPtr	end;			/* exclusive, InvalidXLogRecPtr means infinity */
+	TLHashState	hstate;			/* hash state of TLI's last completed segment at switchpoint */
 } TimeLineHistoryEntry;
 
 extern List *readTimeLineHistory(TimeLineID targetTLI);
 extern bool existsTimeLineHistory(TimeLineID probeTLI);
 extern TimeLineID findNewestTimeLine(TimeLineID startTLI);
 extern void writeTimeLineHistory(TimeLineID newTLI, TimeLineID parentTLI,
-								 XLogRecPtr switchpoint, const char *reason);
+								 XLogRecPtr switchpoint, const char *reason,
+								 TLHashState *hstate);
 extern void writeTimeLineHistoryFile(TimeLineID tli, const char *content, size_t size);
 extern void restoreTimeLineHistoryFiles(TimeLineID begin, TimeLineID end);
 extern bool tliInHistory(TimeLineID tli, List *expectedTLEs);
diff --git a/src/include/access/timeline_hash.h b/src/include/access/timeline_hash.h
new file mode 100644
index 0000000..7b7fa9c
--- /dev/null
+++ b/src/include/access/timeline_hash.h
@@ -0,0 +1,47 @@
+/*
+ * timeline_hash.h
+ *
+ * Definitions for timeline hash state.
+ *
+ * Portions Copyright (c) 1996-2026, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ * src/include/access/timeline.h
+ */
+#ifndef TLHASHST_H
+#define TLHASHST_H
+
+/*
+ * NOTE:
+ * TLHashState is a chained hash state of WAL bytes in current timeline.
+ * It is consistent across nodes that receive identical WAL stream
+ * (replication or archive). It is NOT guaranteed to match across
+ * independent primary nodes, even if they execute similar SQL queries,
+ * because that execution is non-deterministic.
+ *
+ * 'hash'	is hash value, initialized with TLHASHST_SEED;
+ * 'segno'	represent currently hashed segment file number;
+ * 'tli'	represent currently hashed segment file timeline;
+ * 'flags'	represent different hash states.
+ */
+typedef struct {
+	uint64		hash;
+	XLogSegNo	segno;
+	TimeLineID	tli;
+	uint16		flags;
+} TLHashState;
+
+/*
+ * Determines that hash was updated from the start of timeline.
+ */
+#define TLHASHST_FIELDS			4
+#define TLHASHST_SEED			UINT64CONST(0)
+#define TLHASHST_INIT_FLAG		0x0001
+#define TLHASHST_INITIAL		((TLHashState){.hash = TLHASHST_SEED,.segno = 0,.tli = 0,.flags = 0})
+#define TLHASHST_PRINT_ARGS(hs)	(StaticAssertVariableIsOfTypeMacro((hs), TLHashState),(uint64)((hs).hash)),\
+								((uint64)((hs).segno)),((uint32)((hs).tli)),((uint16)((hs).flags))
+#define TLHASHST_SCANF_ARGS(hs)	(StaticAssertVariableIsOfTypeMacro((hs), TLHashState),(uint64*)(&(hs).hash)),\
+								((uint64*)(&(hs).segno)),((uint32*)(&(hs).tli)),((uint16*)(&(hs).flags))
+#define TLHASHST_FORMAT			UINT64_HEX_FORMAT ":" UINT64_FORMAT ":%" PRIu32 ":%" PRIu16
+
+#endif
diff --git a/src/include/access/xlog.h b/src/include/access/xlog.h
index 338d68d..a855f5b 100644
--- a/src/include/access/xlog.h
+++ b/src/include/access/xlog.h
@@ -18,6 +18,7 @@
 #include "lib/stringinfo.h"
 #include "nodes/pg_list.h"
 
+#include "access/timeline_hash.h" /* for TLHashState */
 
 /* Sync methods */
 enum WalSyncMethod
@@ -354,4 +355,11 @@ extern SessionBackupState get_backup_status(void);
 /* files to signal promotion to primary */
 #define PROMOTE_SIGNAL_FILE		"promote"
 
+extern bool XLogHashIsSet(TLHashState hstate);
+extern void XLogResetCurrentHash(void);
+extern bool XLogUpdateCurrentHash(XLogSegNo segno, TimeLineID tli);
+extern uint64 XLogHashSegment(XLogSegNo segno, TimeLineID tli, uint64 seed);
+uint64 GetCurrentTimelineHash(TLHashState *hstate);
+void SetCurrentTimelineHash(TLHashState *hstate);
+
 #endif							/* XLOG_H */
diff --git a/src/include/c.h b/src/include/c.h
index 20cfbac..69ab190 100644
--- a/src/include/c.h
+++ b/src/include/c.h
@@ -694,6 +694,8 @@ typedef uint64_t uint64;
 #define UINT64_FORMAT "%" PRIu64
 #define OID8_FORMAT "%" PRIu64
 
+#define UINT64_HEX_FORMAT "%016" PRIx64
+
 /*
  * 128-bit signed and unsigned integers
  *		There currently is only limited support for such types.
diff --git a/src/include/catalog/pg_control.h b/src/include/catalog/pg_control.h
index 80b3a73..e3f02b4 100644
--- a/src/include/catalog/pg_control.h
+++ b/src/include/catalog/pg_control.h
@@ -20,9 +20,10 @@
 #include "pgtime.h"				/* for pg_time_t */
 #include "port/pg_crc32c.h"
 
+#include "access/timeline_hash.h"
 
 /* Version identifier for this pg_control format */
-#define PG_CONTROL_VERSION	1902
+#define PG_CONTROL_VERSION	1903
 
 /* Nonce key length, see below */
 #define MOCK_AUTH_NONCE_LEN		32
@@ -244,6 +245,9 @@ typedef struct ControlFileData
 	 */
 	char		mock_authentication_nonce[MOCK_AUTH_NONCE_LEN];
 
+	/* Chained hash state of WAL stream of bytes in current timeline */
+	TLHashState hstate;
+
 	/* CRC of all above ... MUST BE LAST! */
 	pg_crc32c	crc;
 } ControlFileData;
diff --git a/src/include/catalog/pg_proc.dat b/src/include/catalog/pg_proc.dat
index 66c3c9a..eea53ce 100644
--- a/src/include/catalog/pg_proc.dat
+++ b/src/include/catalog/pg_proc.dat
@@ -12454,6 +12454,15 @@
   proargnames => '{max_data_alignment,database_block_size,blocks_per_segment,wal_block_size,bytes_per_wal_segment,max_identifier_length,max_index_columns,max_toast_chunk_size,large_object_chunk_size,float8_pass_by_value,data_page_checksum_version,default_char_signedness}',
   prosrc => 'pg_control_init' },
 
+{ oid => '9286',
+  descr => 'pg_controldata hash state information as a function',
+  proname => 'pg_control_hash', provolatile => 'v', prorettype => 'record',
+  proargtypes => '',
+  proallargtypes => '{text,int4,int8,bool}',
+  proargmodes => '{o,o,o,o}',
+  proargnames => '{timeline_hash,hash_tli,hash_segno,init_flag}',
+  prosrc => 'pg_control_hash' },
+
 # subscripting support for built-in types
 { oid => '6179', descr => 'standard array subscripting support',
   proname => 'array_subscript_handler',
